### ISSUE TITLE
Catch 22 With Upgrade

### ISSUE THREAD

**Issue by anthony-sarkis** [helm/helm#10963]
### Situation
The K8s cluster was upgraded. When deployment was attempted it got stuck in a [catch-22](https://en.wikipedia.org/wiki/Catch-22_(logic)) situation. 

```
The 1st attempt to deploy it using the existing helm chart resulted in:
UPGRADE FAILED: unable to recognize "": no matches for kind "Ingress" in version "extensions/v1beta1"
```
In order to deploy the helm chart, the ingress had to be updated to the new networking API (`extensions/v1beta1` to `networking.k8s.io/v1`. However, when this was completed and all little errors sorted, we now ran into: 
```
Error: UPGRADE FAILED: current release manifest contains removed kubernetes api(s) for this kubernetes version and it is therefore unable to build the kubernetes objects for performing the diff. error from kubernetes: unable to recognize “”:no matches for kind “Ingress” in version “"extensions/v1beta1"”
```
Hence the catch 22, since we can’t deploy with the old chart, and we can’t deploy with the new chart.

We worked around this by doing uninstall, which was able to swallowed the error, and re-install.

However, prior to that solution (or becuase of it, still analyzing) while attempting to use the referenced manual scripts, the k8s ingress got into a broken state.

While whatever manual work is our own responsibility, the deprecated API situation seems common enough
that I believe it should be handled as a 1st class citizen in helm. 

I am writing this with the intent of confirming how we can avoid it in the immediate future and trying to offer some thoughts on how to improve it from helms perspective.

### Going forward

We have these options it appears

1) An expectation that we (end user) monitor deprecated APIs, and that we do not upgrade K8s until we have verified it will not break Helm.
2) https://github.com/helm/helm-mapkubeapis
3) ??? Please advise

### Thoughts

1) It does not feel reasonable that helm should allow a user to get into this catch 22 state. To me:

Reasonable: Blocking becuase of _lacking_ APIs in version of k8s (ok I get it if x doesn't exist...)
Unreasonable: Blocking becuase of _legacy_ API versions in k8s (doesn't make sense)

Speaking as a user the wish is: "-auto-handle-legacy-apis".
I get the impression that this would be hard?
But the confusion is that it was still possible to do it by uninstalling  and then re-running the install,
so it naively appears like something similar already exists? 

2) If this is a k8s issue, then can you please tell me what I should raise there so as to not give up on improving this?

3) We usually use helm as part of a CI/CD process. Switching from that, to a manual maintenance script is not good. https://github.com/helm/helm-mapkubeapis seems to make some progress on this, but it still feels like a lot of manual stuff. 
Is there anyway there can be some type of default, or 80% use case wrapper? Could this be integrated directly into helm eg the option above with the auto delete legacy? 

4) It sounds like the support target was something around 1 year
https://helm.sh/blog/helm-2-becomes-unsupported/
As what is hopefully a flattering comparison, postgres offers around 5 years it appears https://www.postgresql.org/support/versioning/ 
For such a major breaking change as what caused this catch-22 situation it seems as if our setup wasn't "that old" that it should cause such issues

5) The docs here https://helm.sh/docs/topics/kubernetes_apis/
provide a minimal maintenance script which, while I appreciate it is better then nothing, speaking as an end user does not feel like an appropriate solution. To me this breaks "trust" in helm as a frontend. It's nearly like asking a non-technical user to just "compile from source". If this was some extreme edge case I would understand it better, but as far as I can tell from searching it's a common issue that people run into so I would appreciate if it was better supported. 

As another analogy it almost feels like offering an app upgrade without the database migration script. Since the migration script is _required_ for the upgrade to succeed, it should be part of the default upgrade flow.

If Helm was a casual side project obviously my entire issue would be probably asking too much.
As a major project with the intent that "everything runs on helm/k8s", this feels like something that has a lot of room for improvement, and is central enough it should be a priority.

In summary, a main use case of helm is continual automatic app deployment, and having something that can get into this type of situation is bad. It seems like there are already a number of solutions in the works, and I tried to add some ideas and feedback here.

**Comment by bacongobbler** [comment]
An understanding of how Helm and Kubernetes handles releases may help in explaining why what you're asking for is difficult.

Helm follows semantic versioning. That means feature additions go in at minor versions and bug fixes go in at patch versions. When we break backwards compatibility, we introduce a new major version. The removal of Tiller in Helm 3 is one example of that.

On the other hand, Kubernetes does _not_ follow semantic versioning. The "extensions/v1beta1" Ingress API (as well as many others, including PetSets, Deployments, ReplicationControllers, CRDs, etc.) can and will be removed in minor version increments. This breaks the basic principles of semantic versioning, where clients can reasonably upgrade from one minor release to the next without expecting any breaking changes.

Forcing resources to migrate from one version to the next is by definition a breaking change. Having Helm do this on behalf of the user could result in catastrophic issues, as the "shape" of the object may change between versions. In some instances, APIs are replaced with an entirely different object (as seen in the past with StatefulSets - once known as PetSets). In many cases, a migration path from one version to the next may not map 1:1, as the semantics of the object can change between versions.

This is why we have been recommending the use of the [helm-mapkubeapis](https://github.com/helm/helm-mapkubeapis) plugin to help with this migration. It updates in-place Helm release metadata that contains deprecated or removed Kubernetes APIs to a new instance with supported Kubernetes APIs. This is provided as a separate plugin so that users are explicitly made aware of this migration behaviour as they migrate their APIs from one version to the next.

I understand the inconvenience here. Kubernetes changed an API in a breaking way. We try to shield from that and sometimes we can accomplish this in different ways. But, in this instance we are not able to. If you have any suggestions how we can better accommodate these workflows, [the contributing guide](https://github.com/helm/helm/blob/main/CONTRIBUTING.md#proposing-an-idea) will help you get started.

Hope this helps.

**Comment by bacongobbler** [comment]
Closing this issue as answered. Please feel free to comment here and let us know if you have further questions. Thanks!
